File: cpp/tct.cpp
#include <iostream>
#include <cmath>
#include <vector>

// Function to calculate standard deviation from a 1D array
template<typename T>
T stddev(const std::vector<T>& array) {
    T bar = 0;
    T sm = 0;
    for (const T& value : array) {
        bar += value;
    }
    int q = array.size();
    bar /= q;
    for (const T& value : array) {
        sm += std::pow(value - bar, 2);
    }
    return std::sqrt(sm / q);
}

// Function to convert stable angle number to ncbb
template<typename T>
T angle2ncbb(T angle_number) {
    if (angle_number < 1.0) {
        return 0.0;
    } else if (angle_number < 3.0) {
        return 1.0;
    } else if (angle_number < 6.0) {
        return 3.0;
    } else if (angle_number < 10.0) {
        return 5.0;
    } else if (angle_number < 15.0) {
        return 7.0;
    } else if (angle_number < 21.0) {
        return 9.0;
    } else if (angle_number < 28.0) {
        return 11.0;
    } else {
        return 13.0;
    }
}

void tct_calculate(int n_atom, const std::vector<std::vector<std::vector<double>>&> vectors,
    std::vector<std::vector<std::vector<double>>&> bond_length,
    std::vector<std::vector<std::vector<double>>&> bond_angle, const std::vector<int>& n_neighbor,
    const std::vector<std::vector<int>>& neighbor, std::vector<std::vector<int>>& length_header,
    std::vector<std::vector<std::vector<int>>&> angle_header, std::vector<std::vector<std::vector<int>>&> counter,
    std::vector<std::vector<double>>& constrain) {

    const double pi = 3.141592653589793;
    const double r2d = 180.0 / pi;

    for (int i = 0; i < n_atom; i++) {
        for (int l = 19; l >= 2; l--) {
            for (int j = 0; j < 10; j++) {
                bond_length[i][j][l] = bond_length[i][j][l - 1];
            }
        }

        for (int l = 19; l >= 2; l--) {
            for (int j = 0; j < 30; j++) {
                bond_angle[i][j][l] = bond_angle[i][j][l - 1];
            }
        }

        for (int j = 0; j < 10; j++) {
            if (counter[i][j][1] == 0 && length_header[i][j] > 0) {
                for (int l = j; l < 9; l++) {
                    for (int k = 0; k < 2; k++) {
                        if (angle_header[i][l][k] == length_header[i][j]) {
                            for (int m = 0; m < 30; m++) {
                                angle_header[i][l][k] = angle_header[i][l + 1][k];
                                bond_angle[i][l][k] = bond_angle[i][l + 1][k];
                            }
                            if (l < 29) {
                                l--;
                            }
                            break;
                        }
                    }
                }
                length_header[i][j] = 0;
                for (int l = j; l < 9; l++) {
                    for (int k = 0; k < 2; k++) {
                        angle_header[i][l][k] = angle_header[i][l + 1][k];
                        bond_angle[i][l][k] = bond_angle[i][l + 1][k];
                    }
                }
                angle_header[i][29][0] = 0;
                angle_header[i][29][1] = 0;
                bond_angle[i][29][0] = 0;
                bond_angle[i][29][1] = 0;
                j--;
            }
        }

        for (int j = 0; j < 10; j++) {
            if (counter[i][j][1] < 20) {
                counter[i][j][1] = 0;
            }
            if (counter[i][j][2] > 0) {
                counter[i][j][2]--;
            }
        }

        for (int j = 0; j < 10; j++) {
            int length = neighbor[i][j];
            double length_val = 0.0;
            for (int m = 0; m < 3; m++) {
                length_val += vectors[i][j][m] * vectors[i][j][m];
            }
            length_val = std::sqrt(length_val);
            for (int l = 0; l < 10; l++) {
                if (neighbor[i][j] == length_header[i][l]) {
                    bond_length[i][j][0] = length_val;
                    counter[i][j][0]++;
                    counter[i][j][1] = 20;
                    break;
                }
                else if (neighbor[i][j] < length_header[i][l]) {
                    for (int k = 9; k >= l; k--) {
                        length_header[i][k + 1] = length_header[i][k];
                        for (int m = 0; m < 20; m++) {
                            bond_length[i][j][m + 1] = bond_length[i][j][m];
                        }
                    }
                    length_header[i][l] = neighbor[i][j];
                    bond_length[i][j][0] = length_val;
                    counter[i][j][0] = 1;
                    counter[i][j][1] = 20;
                    break;
                }
                else if (length_header[i][l] == 0) {
                    length_header[i][l] = neighbor[i][j];
                    bond_length[i][j][0] = length_val;
                    counter[i][j][0]++;
                    counter[i][j][1] = 20;
                    break;
                }
            }
        }

        for (int j = 0; j < n_neighbor[i]; j++) {
            double length_a = 0.0;
            for (int m = 0; m < 3; m++) {
                length_a += vectors[i][j][m] * vectors[i][j][m];
            }
            length_a = std::sqrt(length_a);

            for (int l = j + 1; l < n_neighbor[i]; l++) {
                double angle = r2d * std::acos(
                    (vectors[i][j][0] * vectors[i][l][0] + vectors[i][j][1] * vectors[i][l][1] +
                    vectors[i][j][2] * vectors[i][l][2]) /
                    (length_a * norm2(vectors[i][l])));

                for (int k = 0; k < 30; k++) {
                    if (length_header[i][k] == length_a && angle_header[i][k][0] == neighbor[i][j] && angle_header[i][k][1] == neighbor[i][l]) {
                        bond_angle[i][k][0] = angle;
                        break;
                    }
                    else if (length_a < length_header[i][k]) {
                        for (int m = 29; m >= k; m--) {
                            angle_header[i][m + 1][0] = angle_header[i][m][0];
                            angle_header[i][m + 1][1] = angle_header[i][m][1];
                            for (int n = 0; n < 20; n++) {
                                bond_angle[i][m + 1][n] = bond_angle[i][m][n];
                            }
                        }
                        angle_header[i][k][0] = length_a;
                        angle_header[i][k][1] = neighbor[i][j];
                        bond_angle[i][k][0] = angle;
                        break;
                    }
                    else if (length_header[i][k] == 0) {
                        angle_header[i][k][0] = length_a;
                        angle_header[i][k][1] = neighbor[i][j];
                        bond_angle[i][k][0] = angle;
                        break;
                    }
                }
            }
        }

        int tm = 0;
        int tm_list[10] = {0};

        for (int j = 0; j < 10; j++) {
            if (counter[i][j][0] >= 20 && counter[i][j][1] == 20) {
                tm++;
                tm_list[tm - 1] = length_header[i][j];
                double tct = stddev(bond_length[i][j]);
                if (tct < 1.0) {
                    constrain[i][0] += 1.0;
                }
                for (int l = j + 1; l < 10; l++) {
                    if (counter[i][l][0] >= 20 && counter[i][l][1] == 20) {
                        for (int k = 0; k < 30; k++) {
                            if (angle_header[i][k][0] == length_header[i][j] && angle_header[i][k][1] == length_header[i][l]) {
                                double tct = stddev(bond_angle[i][k]);
                                if (tct < 15.0) {
                                    constrain[i][1] += 1.0;
                                }
                                break;
                            }
                        }
                    }
                }
            }
        }

        constrain[i][0] = 0.5 * constrain[i][0];

        if (constrain[i][1] >= 1.0) {
            constrain[i][1] = angle2ncbb(constrain[i][1]);
        }
    }
}

// int main() {
//     // Example usage of tct_calculate function with appropriate data structures
//     int n_atom = 100; // Replace with your actual number of atoms
//     std::vector<std::vector<std::vector<double>>> vectors(n_atom, std::vector<std::vector<double>>(10, std::vector<double>(3, 0.0)));
//     std::vector<std::vector<std::vector<double>>> bond_length(n_atom, std::vector<std::vector<double>>(10, std::vector<double>(20, 0.0)));
//     std::vector<std::vector<std::vector<double>>> bond_angle(n_atom, std::vector<std::vector<double>>(30, std::vector<double>(20, 0.0)));
//     std::vector<int> n_neighbor(n_atom, 10); // Replace with your actual neighbor data
//     std::vector<std::vector<int>> neighbor(n_atom, std::vector<int>(10, 0)); // Replace with your actual neighbor data
//     std::vector<std::vector<int>> length_header(n_atom, std::vector<int>(10, 0));
//     std::vector<std::vector<std::vector<int>>> angle_header(n_atom, std::vector<std::vector<int>>(30, std::vector<int>(2, 0)));
//     std::vector<std::vector<std::vector<int>>> counter(n_atom, std::vector<std::vector<int>>(10, std::vector<int>(2, 0)));
//     std::vector<std::vector<double>> constrain(n_atom, std::vector<double>(2, 0.0));
//
//     tct_calculate(n_atom, vectors, bond_length, bond_angle, n_neighbor, neighbor, length_header, angle_header, counter, constrain);
//
//     // Output the results in the 'constrain' vector
//     for (int i = 0; i < n_atom; i++) {
//         std::cout << "Atom " << i << ": Constraint 1 = " << constrain[i][0] << ", Constraint 2 = " << constrain[i][1] << std::endl;
//     }
//
//     return 0;
// }

void poly_volume(int n_atom, std::vector<int>& n_neigh, std::vector<std::vector<std::vector<double>>>& delta, int ctype, int n_ctype, std::vector<double>& hull_v, std::vector<double>& std_volume) {
    int dp = 15;
    std::vector<std::vector<double>> edge(n_atom, std::vector<double>(20, 0.0));

    double vfactor = 6.0 * std::sqrt(2.0);
    int c = 0;

    for (int n = 0; n < n_atom; ++n) {
        if (ptype[n] == ctype && n_neigh[n] == 4) {
            int k = 0;
            for (int i = 0; i < n_neigh[n] - 1; ++i) {
                for (int j = i + 1; j < n_neigh[n]; ++j) {
                    double norm = 0.0;
                    for (int d = 0; d < 3; ++d) {
                        double diff = delta[n][j][d] - delta[n][i][d];
                        norm += diff * diff;
                    }
                    edge[n][k] = std::sqrt(norm);
                    k++;
                }
            }

            double mean = 0.0;
            for (int i = 0; i < k; ++i) {
                mean += edge[n][i];
            }
            mean /= k;

            std_volume[c] = hull_v[c] / std::pow(mean, 3) * vfactor;
            c++;
        }
    }
}

// int main() {
//     // Example usage
//     int n_atom = 10;
//     std::vector<int> n_neigh(n_atom, 4);
//     std::vector<std::vector<std::vector<double>>> delta(n_atom, std::vector<std::vector<double>>(10, std::vector<double>(3, 0.0)));
//     int ctype = 1;
//     int n_ctype = 2;
//     std::vector<double> hull_v(n_ctype, 0.0);
//     std::vector<double> std_volume(n_ctype, 0.0);
//
//     poly_volume(n_atom, n_neigh, delta, ctype, n_ctype, hull_v, std_volume);
//
//     return 0;
// }

File: src/tct.f90
module tct
  use precision
  use io, n_atom => natom
  use nf
  implicit none
  save
  public
  real(dp), allocatable :: constrain(:,:)
  
  contains

subroutine tct_calculate
! Constraint is calculated as follow:
! 1. For each atom, 2 lists are created for bond lengths and bond angles correspond
!   to it. It is a n*ndev list for each atom, ndev is the number of frames will be
!   used to calculate standard deviation.
! 2. When having enough data, which means a bond is stable and unbroken for enough
!   time, standard deviation of its corralated angles are calculated.
! 3. If standard deviation is lower than given threshold, it is considered as an
!   active constraint.

  implicit none

  integer(4) :: i,j,k,l,topa,topb,tm,tm_list(10)
  integer(4) :: ndev=20  ! frame number to calculate a statistical result
  real(dp) :: thv=15.   ! threshold value for constraint determine
  real(dp) :: tct,angle,length
  real(dp), parameter :: pi = 3.141592653589793 ,r2d = 180.0/pi

  real(dp), allocatable :: bond_length(:,:,:), bond_angle(:,:,:)
  integer(4), allocatable :: length_header(:,:), angle_header(:,:,:), counter(:,:,:)


  call neighbor_finder_d

  allocate(bond_length(n_atom,10,ndev))
  allocate(bond_angle(n_atom,30,ndev))
  allocate(length_header(n_atom,10))
  allocate(angle_header(n_atom,30,2))
  allocate(counter(n_atom,10,2))

  allocate(constrain(n_atom,2))

  do i =1, n_atom
    bond_length(i,:,2:ndev)=bond_length(i,:,:ndev-1)  !move bond_length
    bond_angle(i,:,2:ndev)=bond_angle(i,:,:ndev-1)  !move bond angle
    do j=1,10
      if (counter(i,j,2)/=ndev) then
        counter(i,j,1)=0
      endif
      if (counter(i,j,2)>0) then
        counter(i,j,2)=counter(i,j,2)-1
      endif
    enddo
    j = 1
    do !clear space for bond length
      if (counter(i,j,1)==0 .and. length_header(i,j) > 0) then
        bond_length(i,j:-1,:)=bond_length(i,j+1:,:)
        bond_length(i,10,:)=0
        l=1
        do !l = 1,20
          do k = 1,2
            !clear space for bond angle
            if (angle_header(i,l,k)==length_header(i,j)) then
              bond_angle(i,l:-1,:)=bond_angle(i,l+1:,:)
              bond_angle(i,30,:)=0
              angle_header(i,l:-1,:)=angle_header(i,l+1:,:)
              angle_header(i,30,:)=0
              if (l<29) then
              l=l-1
              endif
              exit
            endif
          enddo
          l=l+1
          ! if (l==30) then
          !   print *, i, j, l
          ! endif
          if (l>30) exit
        enddo
        length_header(i,j:-1)=length_header(i,j+1:)
        length_header(i,10)=0
        counter(i,j:-1,:)=counter(i,j+1:,:)
        counter(i,10,:)=0
        j=j-1
      end if
      j=j+1
      if (j>10) exit
    enddo

    do j = 1,n_neighbor(i)
      length = norm2(delta(i,j,:))
      do l = 1,10
        if (neighbor(i,j)==length_header(i,l)) then
          bond_length(i,l,1)=length
          counter(i,l,1)=counter(i,l,1)+1
          counter(i,l,2)=ndev
          exit  !exit if this bond length is stored
        elseif (neighbor(i,j)<length_header(i,l)) then
          length_header(i,l+1:)=length_header(i,l:-1)
          length_header(i,l)=neighbor(i,j)
          bond_length(i,l+1:,:)=bond_length(i,l:-1,:)
          bond_length(i,l,1)=length
          counter(i,l+1:,:)=counter(i,l:-1,:)
          counter(i,l,1)=1
          counter(i,l,2)=ndev
          exit
        elseif (length_header(i,l)==0) then
          length_header(i,l+1:)=length_header(i,l:-1)
          length_header(i,l)=neighbor(i,j)
          bond_length(i,l,1)=length
          counter(i,l,1)=counter(i,l,1)+1
          counter(i,l,2)=ndev
          exit
        endif
        if (l==10) then
          print *, 'Can"t find a place to insert bond length data.'
        endif
      enddo
      do l = j+1,n_neighbor(i)
        topa=neighbor(i,j)
        topb=neighbor(i,l)
        angle= r2d*acos((delta(i,j,1)*delta(i,l,1)+delta(i,j,2)*&
        &delta(i,l,2)+delta(i,j,3)*delta(i,l,3))/(norm2(delta(i,j,:))*&
        &norm2(delta(i,l,:))))

        do k = 1,30
          if (topa==angle_header(i,k,1) .and. topb==angle_header(i,k,2)) then
            bond_angle(i,k+1:,:)=bond_angle(i,k:-1,:)
            bond_angle(i,k,1)=angle
            exit
          elseif (topa<angle_header(i,k,1)) then
            angle_header(i,k+1:,:)=angle_header(i,k:-1,:)
            angle_header(i,k,1)=topa
            angle_header(i,k,2)=topb
            bond_angle(i,k+1:,:)=bond_angle(i,k:-1,:)
            bond_angle(i,k,1)=angle
            exit
          elseif (topa==angle_header(i,k,1) .and. topb<angle_header(i,k,2)) then
            angle_header(i,k+1:,:)=angle_header(i,k:-1,:)
            angle_header(i,k,1)=topa
            angle_header(i,k,2)=topb
            bond_angle(i,k+1:,:)=bond_angle(i,k:-1,:)
            bond_angle(i,k,1)=angle
            exit
          elseif (topa>angle_header(i,k,1) .and. k>1 .and. topa<=angle_header(i,k-1,1)) then
            angle_header(i,k+1:,:)=angle_header(i,k:-1,:)
            angle_header(i,k,1)=topa
            angle_header(i,k,2)=topb
            bond_angle(i,k+1:,:)=bond_angle(i,k:-1,:)
            bond_angle(i,k,1)=angle
            exit
          elseif (angle_header(i,k,1)==0) then
            angle_header(i,k+1:,:)=angle_header(i,k:-1,:)
            angle_header(i,k,1)=topa
            angle_header(i,k,2)=topb
            bond_angle(i,k+1:,:)=bond_angle(i,k:-1,:)
            bond_angle(i,k,1)=angle
            exit
          endif
          if (k==30) then
            print *, 'Can"t find a place to insert bond angle data.'
          endif
        enddo
      enddo
    enddo

    tm=0
    tm_list=0
    do j=1,10
      if (counter(i,j,1)>=ndev .and. counter(i,j,2)==ndev) then
        tm=tm+1
        tm_list(tm)=length_header(i,j)
        !stddva of bond length
        tct=stddev(bond_length(i,j,:))
        if (tct<1.) then
          constrain(i,1) = constrain(i,1)+1.
        endif
        do l=j+1,10
          if (counter(i,l,1)>=ndev .and. counter(i,l,2)==ndev) then
            do k=1,30
              if (angle_header(i,k,1)==length_header(i,j) .and. &
              &angle_header(i,k,2)==length_header(i,l)) then
                tct = stddev(bond_angle(i,k,:))
                if (tct<thv) then   !here is the threshold value that need to be modified
                  constrain(i,2) = constrain(i,2)+1.
                endif
                exit
              endif
            enddo
          endif
        enddo
      endif
    enddo
    constrain(i,1)=0.5*constrain(i,1)
    if (constrain(i,2)>=1.) then
      constrain(i,2)=angle2ncbb(constrain(i,2))
      ! sqrt(8.*constrain(i,2)+1.)-2.
    endif
  enddo

  ! print *, bond_length(886315,2,:)
  ! print *, delta(886315,2,:)

  deallocate(bond_length)
  deallocate(bond_angle)
  deallocate(length_header)
  deallocate(angle_header)
  deallocate(counter)

  return
  contains
  function stddev(array)
    ! a function to calculate standard deviation from a 1D array.
    implicit none
    real(dp), intent(in) :: array(:)
    real(dp) :: bar, sm, stddev
    integer(4) :: p, q
    sm=0.
    q=size(array)
    bar = sum(array)/q
    do p = 1,q
      sm = sm+(array(p)-bar)**2.
    enddo
    stddev=sqrt(sm/q)
    return
  end function stddev

  function angle2ncbb(angle_number)
    ! a step function to get ncbb from stable angle number, is a little bit ambigious.
    implicit none
    real(dp), intent(in) :: angle_number
    real(dp) :: angle2ncbb
    if (angle_number < 1.) then
      angle2ncbb = 0.
    else if (angle_number < 3.) then
      angle2ncbb = 1.
    else if (angle_number < 6.) then
      angle2ncbb = 3.
    else if (angle_number < 10.) then
      angle2ncbb = 5.
    else if (angle_number < 15.) then
      angle2ncbb = 7.
    else if (angle_number < 21.) then
      angle2ncbb = 9.
    else if (angle_number < 28.) then
      angle2ncbb = 11.
    else
      angle2ncbb = 13.
    end if
    return
  end function angle2ncbb

end subroutine tct_calculate


subroutine poly_volume(n_atom,n_neigh,delta,ptype,ctype,n_ctype,hull_v,std_volume)
  implicit none
  integer(4), intent(in) :: n_atom,n_neigh(n_atom),ptype(n_atom),ctype,n_ctype
  real(dp), intent(in) :: delta(n_atom,10,3),hull_v(n_ctype)
  real(dp), intent(out) :: std_volume(n_ctype)
  real(dp) :: edge(n_atom,20), vfactor,mean
  integer(4) :: n,k,i,j,c
  vfactor=6.*sqrt(2.)
  c=1
  do n=1,n_atom
    if (ptype(n)==ctype .and. n_neigh(n)==4) then
      k=1
      do i=1,n_neigh(n)-1
        do j=i+1,n_neigh(n)
          edge(n,k)=norm2((delta(n,j,:)-delta(n,i,:)))
          k=k+1
        enddo
      enddo
      mean = sum(edge(n,:k-1))/(k-1)
      std_volume(c)=hull_v(c)/mean**3*vfactor
      c=c+1
    endif
  enddo
end subroutine poly_volume

end module tct
